Introduce reworked API for ADC drivers

anangl opened the pull request:
This replaces #6176 and introduces a new API for ADC drivers basing on the feedback gathered there.
Implementations of drivers for ADC peripherals present in nRF chips are included, as well as updated and extended adc_api test to present the way the API is supposed to be used.

Marked as DNM, since existing ADC drivers (and "grove" sensors) need to be updated for this one to be merged.

List of drivers to be converted
- [x] nrfx
- [x] dw/qmsi
- [x] sam_afec
- [ ] ti_adc108s102
- [x] mcux

Fixes #3980

anangl committed:
drivers: adc: Introduce reworked API

This commit replaces the API for ADC drivers with a reworked one.
As requested in the issue #3980, the adc_enable/adc_disable functions
are removed. Additionaly, some new features are introduced, like:
- asynchronous calls
- configuration of channels
- multi-channel sampling

Common parts of code that are supposed to appear in each implementation
of the driver (like locking, synchronization, triggering of consecutive
samplings) are provided in the "adc_context.h" file to keep consistency
with the SPI driver. Syscalls are no longer present in the API because
the functions starting read requests cannot use them, since they can be
provided with a callback that is executed in the ISR context, and there
is no point in supporting syscalls only for the channels configuration.

"adc_api" test is updated and extended with additional test cases,
with intention to show how the API is supposed to be used.
"adc_simple" test is removed as it does not seem to add much value.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: adc: Add shims for nrfx ADC and SAADC drivers

This commit adds translation layers to make nrfx drivers for the nRF
ADC (nRF51 series) and SAADC (nRF52 series) peripherals accessible via
the Zephyr's API. The SAADC peripheral is accessed using nrfx HAL only
as it turns out that usage of the nrfx driver in this case would be
inconvenient and would unnecessarily complicate the shim.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

tbursztyka commented on include/adc.h:
```diff
@@ -26,121 +26,303 @@ extern "C" {
+	u8_t channel_id   : 5;
+
+	/** Channel type: single-ended or differential. */
+	bool differential : 1;
```
Keep u8_t here, I am not entirely sure compiler will properly pack it with previous byte.

anangl commented:
- refactored the test a bit, added one more test case for asynchronous calls
- added an atomic variable to handle triggering of another sampling while the previous one is still in progress
- added the possibility to perform multiple samplings with zero interval (as fast as possible)

galak commented on drivers/adc/adc_nrfx_adc.c:
```diff
@@ -0,0 +1,269 @@
+		return -EBUSY;
+	}
+
+	IRQ_CONNECT(ADC_IRQn, CONFIG_ADC_0_IRQ_PRI,
```
We should get IRQ info from DT

anangl replied:
done

galak commented on drivers/adc/adc_nrfx_adc.c:
```diff
@@ -0,0 +1,269 @@
+static nrfx_adc_channel_t m_channels[CONFIG_ADC_NRFX_ADC_CHANNEL_COUNT];
+
+
+static int adc_nrfx_channel_setup(struct device *dev,
```
use nrfx_adc_ as a function prefix, its a bit confusing when seeing adc_ and thinking maybe that's an adc core function, and not a driver specific one.

anangl replied:
But the `nrfx_` prefix is used for stuff that comes from nrfx, and in particular all API functions of the nrfx driver for ADC peripheral are prefixed with `nrfx_adc_`. So that would be confusing as well. On the other hand, `adc_nrfx_` is consistent with the scheme we use for naming implementation files.
And as I see such approach is commonly used. See for example: `spi_dw_transceive`, `i2c_mcux_configure`, `uart_stm32_poll_in`. Thus, I would prefer to stay with this naming.

galak replied:
ok, that makes sense.

galak replied:
I guess my comment is about being consistent in this file, so either everything should be prefixed adc_nrfx_ or not.

galak commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -0,0 +1,374 @@
+	nrf_saadc_int_enable(NRF_SAADC_INT_END);
+	NRFX_IRQ_ENABLE(SAADC_IRQn);
+
+	IRQ_CONNECT(SAADC_IRQn, CONFIG_ADC_0_IRQ_PRI,
```
IRQ num and priority should come from DT.

anangl replied:
done

galak reviewed (changes requested):
Use DT for IRQ info, should have DT nodes for adc controllers.

anangl committed:
dts: nrf: Add ADC nodes and bindings for nRF SoCs

This commit adds ADC nodes to DTS files for nRF SoCs and introduces
corresponding  bindings for these nodes.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: nrf: Enable ADC nodes in DTS for nRF boards

ADC nodes are enabled in DTS for the following nRF development kits:
- nrf51_pca10028
- nrf52_pca10040
- nrf52_pca20020
- nrf52810_pca10040
- nrf52840_pca10056

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
tests: adc_api: Add configurations for nRF5 boards

This adds to the adc_api test board-specific configurations
for the following boards:
- nrf51_pca10028
- nrf52_pca10040
- nrf52840_pca10056

"adc" is indicated as supported in the yaml files for these boards
so that they are included in sanitycheck.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
- rebased on the latest master to get the support for nRF52810 and update its "ADC affected" files in this PR
- added DTS entries as requested by @galak
- minor editorial changes: used `BIT` and `BIT_MASK` macros instead of shifting in `ADC_ACQ_TIME` macros, renamed `ADC_DEV_NAME` TO `ADC_DEVICE_NAME` in the "adc_api" test

anangl commented:
- removed syscalls from the API since read functions cannot use them because of the callback that is executed in the ISR context, and there is a little sense in having syscall only for channel configuration

anangl commented:
Rebased on master to solve a conflict in "boards/arm/nrf52_pca20020/Kconfig.board".

galak commented on boards/arm/nrf51_pca10028/Kconfig.board:
```diff
@@ -6,4 +6,5 @@
 
 config BOARD_NRF51_PCA10028
 	bool "nRF51 PCA10028"
+	select HAS_DTS_ADC
```
select HAS_DTS_ADC, should be with the driver, not the board.

galak commented on boards/arm/nrf52810_pca10040/Kconfig.board:
```diff
@@ -6,5 +6,6 @@
 
 config  BOARD_NRF52810_PCA10040
 	bool "nRF52810 PCA10040"
+	select HAS_DTS_ADC
```
remove this, just do the select in the driver's Kconfig.

galak commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -0,0 +1,374 @@
+	 */
+	channel_id = 0;
+	do {
+		if (channels & 1) {
```
seems like the 1 here could be replaced w/define or some support.  Is this an enable bit?

galak commented on drivers/adc/Kconfig:
```diff
@@ -13,10 +13,16 @@ menuconfig ADC
 	bool
 	prompt "ADC drivers"
```
add `select HAS_DTS_ADC` so that all ADC drivers require DTS support.

anangl replied:
But what would be the point in having `HAS_DTS_ADC` at all then? Right now it only excludes from Kconfig the options providing information that can be taken from DT, and these options are dependent on the same menuconfig entry that would select `HAS_DTS_ADC`. They should be rather removed then.
And this would be inconsistent with what we've got in other drivers. Is this a new approach that we want to take?

anangl replied:
As agreed on the API call, I'll add another commit with this change.

galak reviewed (changes requested):
Some comments on the DTS support and minor comments on code style for the NRF drivers.

anangl commented:
@galak Regarding the "adc_nrfx_" prefix (I don't know why github does not allow me to reply to your comment directly). I decided to keep such prefix only for stuff directly related to the ADC driver API, i.e. the functions implementing the mandatory API and the structure holding the pointers to these functions. Using the prefix for all static functions in the implementation file seems to be impractical, and on the other hand, getting rid of the prefix completely would lead to having functions "read" and "read_async". That wouldn't look good either.

galak commented on dts/bindings/adc/nordic,nrf-adc.yaml:
```diff
@@ -0,0 +1,35 @@
+#
```
Put this file in dts/bindings/iio/adc

anangl replied:
done

galak reviewed (dismissed):
Move dts/bindings/adc/nordic,nrf-adc.yaml to dts/bindings/iio/adc/nordic,nrf-adc.yaml

anangl committed:
drivers: adc: Select HAS_DTS_ADC for all ADC drivers

To require relevant DTS entries on all platforms providing the ADC
driver.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
- added/polished some comments in "adc.h"

anangl commented:
- corrected the contents of "adc_context.h", "test_adc.c", and "adc/Kconfig" to get rid of one error and a few warnings reported by checkpatch

anangl commented:
- rebased on master to resolve the conflict in _arch/arm/soc/nordic_nrf/nrf51/dts.fixup_

anangl commented on include/adc.h:
```diff
@@ -26,121 +26,314 @@ extern "C" {
  */
-struct adc_seq_entry {
-	/** Clock ticks delay before sampling the ADC. */
-	s32_t sampling_delay;
```
The API is intended to be generic, so it cannot directly expose all features of particular hardware. If I correctly understand the purpose of this sampling delay feature, you can use it to realize the intervals between samplings that can by specified in [sequence options](https://github.com/zephyrproject-rtos/zephyr/pull/7691/files#diff-895d709eadff47267669f2993f5ff80eR157), instead of involving a timer for this.

punitvara replied:
Ok this makes sense.

anangl commented on include/adc.h:
```diff
@@ -26,121 +26,314 @@ extern "C" {
+	size_t buffer_size;
 
-	/* Number of entries in the sequence entry array. */
-	u8_t num_entries;
```
You can select multiple channels to be sampled in a sequence. It is done by setting corresponding bits in the `channels` field of the `adc_sequence` structure. By using the `adc_sequence_options` and its `extra_samplings` field you can also request given sequence to be performed multiple times.

punitvara replied:
OK so I understand previously what number entries used to represent is now extra_sampling

anangl replied:
Almost. `extra_samplings` is the number of additional samplings, as one is performed always. So, the previous `num_entries` will be equivalent to `1 + extra_samplings` now. And `extra_sampling`, as the whole `adc_sequence_options` containing it, is optional and may not appear at all.

punitvara replied:
Ok.

anangl commented:
- described more precisely the purpose of the "channel_id" field to avoid confusion regarding the way ADC inputs should be selected
- made the `ADC_ACTION_CONTINUE` the default action taken in `adc_context_on_sampling_done`, just in case the callback returns some weird value
- corrected a typo in one comment in `adc_context.h`

punitvara commented:
I am still not sure about `channel_id` and `channels` field. User provides channels in adc_sequence which indicated driver that use want channel (i.e Channel X and Channel Y) according to bit map passed. And it has to be set in hardware to activate that respective channel. Now what `channel_id `represent ? what I only understood from your comment is value of `channel_id `doesn't need to be written on hardware register but then what is importance or use of that field. Can you please elaborate ?

punitvara commented on include/adc.h:
```diff
@@ -26,121 +26,324 @@ extern "C" {
+	u8_t differential : 1;
+
+	/** Positive ADC input. */
+	u8_t input_positive;
```
What user populate in this variable ? pin number from which positive input should taken ?

anangl commented:
@punitvara
A channel has to be configured first with the `adc_api_channel_setup` function. In the provided `adc_channel_cfg` structure user assigns an identifier (`channel_id`) for the channel and selects an ADC input (or inputs) to be associated with the channel (by filling the `input_positive` field, and for a differential channel, `input_negative` as well). Then the user can select the channel for a sampling sequence by setting the corresponding bit (`BIT(0)` for `channel_id` 0, `BIT(1)` for `channel_id` 1 and so on) in the `channels` bit-mask.
See how it is done in the ADC API test: [channel 0 defined](https://github.com/zephyrproject-rtos/zephyr/pull/7691/files#diff-64f1acdc193ee0d6e5b80c6b88aeb6e7R54), [channel 2 defined](https://github.com/zephyrproject-rtos/zephyr/pull/7691/files#diff-64f1acdc193ee0d6e5b80c6b88aeb6e7R63), [channel 0 selected](https://github.com/zephyrproject-rtos/zephyr/pull/7691/files#diff-64f1acdc193ee0d6e5b80c6b88aeb6e7R118), [both channels selected](https://github.com/zephyrproject-rtos/zephyr/pull/7691/files#diff-64f1acdc193ee0d6e5b80c6b88aeb6e7R150).
Particular drivers may support only a limited range of channel identifiers, the maximum range allowed by the API is 0-31. I think I must add such statement to the documentation. I see now that currently this is not clear, moreover, [this "(0-31)"](https://github.com/zephyrproject-rtos/zephyr/pull/7691/files#diff-895d709eadff47267669f2993f5ff80eR94) may be interpreted differently, as a requirement for all drivers. I'll correct this.

What is to be written in the `input_positive` and `input_negative` fields depends on the hardware the driver is handling, and consequently is driver dependent. This numbers must identify analog inputs to be used. Usually, only a set of dedicated pins can serve as analog inputs, so pin numbers won't be suitable in this case. And some hardware may allow to select additional inputs (apart from the pin ones), like VDD (SAADC in Nordic chips is such case). For Nordic drivers I used the enumerations we've got defined for our peripherals: [ADC](https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/nordic/nrfx/hal/nrf_adc.h#L100) and [SAADC](https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/nordic/nrfx/hal/nrf_saadc.h#L68) (these are the values that are actually written to the peripheral registers).

punitvara commented:
Thanks for updating comments @anangl If you look at old test case, channel was passed as follow : 
```
#if defined(CONFIG_BOARD_FRDM_K64F)
#define ADC_DEV_NAME    CONFIG_ADC_1_NAME
#define ADC_CHANNEL     14
#elif defined(CONFIG_BOARD_FRDM_KL25Z)
#define ADC_DEV_NAME    CONFIG_ADC_0_NAME
#define ADC_CHANNEL     12
#elif defined(CONFIG_BOARD_FRDM_KW41Z)
#define ADC_DEV_NAME    CONFIG_ADC_0_NAME
#define ADC_CHANNEL     3
#elif defined(CONFIG_BOARD_HEXIWEAR_K64)
#define ADC_DEV_NAME    CONFIG_ADC_0_NAME
#define ADC_CHANNEL     16
#elif defined(CONFIG_BOARD_HEXIWEAR_KW40Z)
#define ADC_DEV_NAME    CONFIG_ADC_0_NAME
#define ADC_CHANNEL     1
#else
#define ADC_DEV_NAME    CONFIG_ADC_0_NAME
#define ADC_CHANNEL     10
#endif
```
In new test case what user will pass `channels` and `channel_id ` ? Consider any of above case i.e. `ADC_CHANNEL 10 `

anangl commented:
If I understand it correctly, this `10` specifies which input should be sampled, so it should be written to `input_positive`. Thus, in the new test you'll need to set
```
#define ADC_CHANNEL_0_INPUT	10
```
in the part customized for your board.

punitvara commented:
Ok `channels` is to use bit mask for multiple channels. 

If user want to get input from channel 10 and 14 then 
`.channels = BIT(10) | BIT(14)
`
Respective bits are set in `u32_t channels` 

 What is use of two different variable to get ADC input `channel_id `and `input_positive`  In old case only `channel_id` was required for same purpose. 

For above comment  if user want to get the input from channel 10, is following correct way to do it ? 

```
. input_positive  = 10,
.channel_id = 10,
.channels = BIT(10)
```

MaureenHelm committed:
adc: Convert mcux adc16 driver to the new adc api

Major rework of the mcux adc16 driver to convert it to the new adc api.
Currently supports a subset of the api features including synchronous
and asynchronous reads, and consecutive reads triggered by the kernel
timer.

Does not yet support some of the channel configuration options such as
gain and reference voltage because the hardware only allows these
options to be configured by peripheral instance rather than by channel.
The values are currently hardcoded in the driver, but in the future we
could introduce some flexibility per instance via device tree
attributes.

Does not yet support consecutive reads triggered by a hardware timer.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
tests: adc_api: Add support for kinetis boards

Adds board-specific macros to enable the adc_api test to run on kinetis
boards. This restores the test to support all the nxp boards that were
supported before the new adc api was introduced.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

punitvara committed:
drivers: grove: Modify light sensor

Replace old ADC apis with new ones.

Signed-off-by: Punit Vara <punit.vara@intel.com>

anangl commented:
From the discussion with @punitvara and from looking at #9386 I see that by changing the meaning of the `channel_id` field (in comparison to the previous API) I introduced some unnecessary confusion. To not make it difficult for people that used the previous API to update the drivers and applications, I decided to add a hidden Kconfig option `ADC_CONFIGURABLE_INPUTS` that can be selected by the drivers that actually need to have configurable analog inputs for channels (like Nordic ones). When this option is not selected (which is the default setting), the `channel_id` field both identifies the channel within API and selects corresponding channel in the hardware (like it was in the previous API).
Hopefully, this will make things clear and will be convenient to use.

I updated the "test_adc" accordingly and improved its "test_adc_repeated_samplings" case so that it presents better the API features (and is more difficult to pass by a non-compliant implementation).

I kept the previous version of my PR in my branch `anangl:adc_api_rework_forced_inputs`, in case it is needed for reference.

punitvara committed:
drivers: grove: Update temperature driver

Replace old ADC driver APIs with new ones.

Signed-off-by: Punit Vara <punit.vara@intel.com>

fallrisk committed:
sam: adc: Updated SAM ADC driver.

Updated to work with new ADC API.

Signed-off-by: Justin Watson <jwatson5@gmail.com>

punitvara commented on tests/drivers/adc/adc_api/src/test_adc.c:
```diff
@@ -8,98 +9,333 @@
+	const struct adc_sequence sequence = {
+		.channels    = BIT(ADC_1ST_CHANNEL_ID),
+		.buffer      = m_sample_buffer,
+		.buffer_size = sizeof(m_sample_buffer),
```
@anangl  `sizeof(m_sample_buffer) ` this returns 4 byte (size of array). Please change it to ` BUFFER_SIZE * sizeof(m_sample_buffer[0])`. We were facing crash while porting designware driver.

MaureenHelm commented:
Rebased to latest master and resolved merge conflicts.

punitvara committed:
drivers: Use Designware driver for sensor subsystem

Remove adc_qmsi_ss to use designware driver for sensor subsystem

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
dts: arc: Use dts tree for designware driver

Get required data from dts tree.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
drivers: adc: Update new ADC APIs

Replace new ADC apis for designware driver

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
tests: adc_api: Add ARC related parameters for ADC

Add ARC related parameters and modified test case to work
this test case on arc sensor subsystem.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
drivers: Kconfig.dw: Remove unnecessary Kconfig

Remove Kconfig which are replaced by dts and unnecessary macros which is
not useful anymore.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
boards: quark_se_c1000_ss_devboard: Enable ADC node

Enable ADC node in dts.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
samples: grove_light: Provide device name from Kconfig

check return value for sample fetch. Conditional printk
added to print adc value.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
samples: grove_temperature: Use device name from Kconfig

Use Kconfig to get device name. Make LIBC conditional to get
output.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
boards: quark_d2000_crb: Remove adc support

Remove ADC support from d2000 as it is not migrated to
new ADC apis.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
tests: adc_api: Remove sample width test scenario

Remove unnecessary sample width scenario.

Signed-off-by: Punit Vara <punit.vara@intel.com>

dbkinder reviewed (approved):
Doc change LGTM. Commit message should be updated (still talks about being marked as DNM).

MaureenHelm commented:
> Commit message should be updated (still talks about being marked as DNM).

Which commit?

mnkp commented:
I am aware that I am very late to the review, unfortunately I was not able to devote much time to the Zephyr project in last months.

The new ADC API is definitely the step in the right direction, however its design has some problems. The main one is the lack of support for triggers. ADC hardware comes in an overwhelming variety of design variants often with a distinctively different architectures. It makes it difficult to design universal API, there are however some fairly common features. One of such features is the concept of a trigger. Once the ADC module is configured - using the nomenclature of this API - the sequence is defined, the measurement can be started by a trigger whereas trigger is typically:
- a software trigger
- a timer trigger
- a pin
- a trigger of some internal trigger subsystem

of those the new API supports software trigger and makes an attempt to support timer trigger. The pin trigger and subsystem trigger is missing. The subsystem trigger is popular on many low power MCUs where it is possible to start ADC measurement on some internal, predefined condition without waking up the CPU.

Unfortunately this API can not be easily extended. The next time a project needs to trigger ADC measurement on the pin level change the ADC API would need to be completely reworked.
